# Question 1. Merging Instacart Data--orders

1.Download the new data set.

2.Unzip the file and move it into your project folder.

3.Import the data set into Jupyter as a new dataframe: df_ords_prior.
    
4.Check the dimensions of the imported dataframe.

5.Merge the dataframe with df_ords dataframe.

6.Use the indicator argument to check whether there was a full match between the two dataframes.

In [6]:
# Import libarires
import pandas as pd
import numpy as np
import os

In [7]:
# Create path
path = r'C:\Users\SuperHome\Documents\Instacart Basket Analysis\02 Data'

In [8]:
# import dataframe -- 'order_products_prior'
df_ords_prior = pd.read_csv(os.path.join(path,'Original Data','order_products_prior.csv'), index_col=False)

In [9]:
# Check dimensions
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [10]:
# Import datadframe -- 'orders'
df_ords = pd.read_csv(os.path.join(path,'Prepared Data','orders_clean.csv'), index_col=False)

In [11]:
# Check dimensions
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  day_of_week  \
0             0           0   2539329        1             1            2   
1             1           1   2398795        1             2            3   
2             2           2    473747        1             3            3   
3             3           3   2254736        1             4            4   
4             4           4    431534        1             5            4   

   order_hour_of_day  days_since_prior_order  
0                  8                     7.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [12]:
df_ord = df_ords.drop(columns = ['Unnamed: 0.1','Unnamed: 0'])

In [13]:
df_ord

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0         2539329        1             1            2                  8   
1         2398795        1             2            3                  7   
2          473747        1             3            3                 12   
3         2254736        1             4            4                  7   
4          431534        1             5            4                 15   
...           ...      ...           ...          ...                ...   
3421078   2266710   206209            10            5                 18   
3421079   1854736   206209            11            4                 10   
3421080    626363   206209            12            1                 12   
3421081   2977660   206209            13            1                 12   
3421082    272231   206209            14            6                 14   

         days_since_prior_order  
0                           7.0  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [14]:
# Check the output
df_ords_prior.shape

(32434489, 4)

Merging those two dataframes

In [15]:
df_merged_ord = df_ord.merge(df_ords_prior, on = 'order_id', indicator = True)

In [16]:
df_merged_ord

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0          2539329        1             1            2                  8   
1          2539329        1             1            2                  8   
2          2539329        1             1            2                  8   
3          2539329        1             1            2                  8   
4          2539329        1             1            2                  8   
...            ...      ...           ...          ...                ...   
32434484   2977660   206209            13            1                 12   
32434485   2977660   206209            13            1                 12   
32434486   2977660   206209            13            1                 12   
32434487   2977660   206209            13            1                 12   
32434488   2977660   206209            13            1                 12   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                            7.0         196                  1          0   
1                            7.0       14084                  2          0   
2                            7.0       12427                  3          0   
3                            7.0       26088                  4          0   
4                            7.0       26405                  5          0   
...                          ...         ...                ...        ...   
32434484                     7.0       14197                  5          1   
32434485                     7.0       38730                  6          0   
32434486                     7.0       31477                  7          0   
32434487                     7.0        6567                  8          0   
32434488                     7.0       22920                  9          0   

         _merge  
0          both  
1          both  
2          both  
3          both  
4          both  
...         ...  
32434484   both  
32434485   both  
32434486   both  
32434487   both  
32434488   both  

[32434489 rows x 10 columns]

In [17]:
df_merged_ord['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

In [18]:
df_merged_test = df_ord.merge(df_ords_prior, on = ['order_id'], how = 'outer', indicator = True)

In [19]:
df_merged_test['_merge'].value_counts()

both          32434489
left_only       206209
right_only           0
Name: _merge, dtype: int64

We actually merged 206209 rows

# Question 2. Exporting Data in Pickle Format

In [20]:
df_merged_ord.to_pickle(os.path.join(path,'Prepared Data','orders_products_combined.pkl'))